src/rhash/hash.c

```diff
@@ -1223,6 +1223,51 @@ static int rc_hash_wii(md5_state_t* md5,
   return 1;
 }
 
+static int rc_hash_wiiware(md5_state_t* md5,
+                           const rc_hash_iterator_t* iterator,
+                           void* file_handle)
+{
+  uint32_t cert_chain_size, ticket_size, tmd_size, data_size;
+  uint32_t tmd_start_addr, hash_size;
+
+  uint8_t quad_buffer[4];
+  uint8_t* buffer;
+
+  /* GetApploaderSize */
+  rc_file_seek(iterator, file_handle, 0x08, SEEK_SET);
+  rc_file_read(iterator, file_handle, quad_buffer, 4);
+  cert_chain_size =
+    ((uint32_t)quad_buffer[0] << 24) | (quad_buffer[1] << 16) | (quad_buffer[2] << 8) | quad_buffer[3];
+  /* Round up to nearest 0x40 */
+  cert_chain_size = ((cert_chain_size - 1) / 0x40 + 1) * 0x40;
+  rc_file_seek(iterator, file_handle, 0x10, SEEK_SET);
+  rc_file_read(iterator, file_handle, quad_buffer, 4);
+  ticket_size =
+    ((uint32_t)quad_buffer[0] << 24) | (quad_buffer[1] << 16) | (quad_buffer[2] << 8) | quad_buffer[3];
+  ticket_size = ((ticket_size - 1) / 0x40 + 1) * 0x40;
+  rc_file_read(iterator, file_handle, quad_buffer, 4);
+  tmd_size =
+    ((uint32_t)quad_buffer[0] << 24) | (quad_buffer[1] << 16) | (quad_buffer[2] << 8) | quad_buffer[3];
+  tmd_size = ((tmd_size - 1) / 0x40 + 1) * 0x40;
+  rc_file_read(iterator, file_handle, quad_buffer, 4);
+  data_size =
+    ((uint32_t)quad_buffer[0] << 24) | (quad_buffer[1] << 16) | (quad_buffer[2] << 8) | quad_buffer[3];
+  data_size = ((data_size - 1) / 0x40 + 1) * 0x40;
+
+  tmd_start_addr = 0x20 + cert_chain_size + ticket_size;
+  hash_size = tmd_size + data_size;
```

Comment:
recommend having an upper limit here in case something that looks like a WiiWare file gets through and the sizes are garbage.
```
if (hash_size > MAX_BUFFER_SIZE)
    hash_size = MAX_BUFFER_SIZE;
```